### PR TITLE
[Qwen3] Qwen3 MoE support 

### PR DESCRIPTION
**Pull request by wwwjn**
As Qwen3 dense model and MoE model share a lot of common parts (eg,  Attention), I added MoE module on top of Qwen3 dense model.

### CODE REVIEW

**Comment by jthomy** [comment]
Great to see! 
I was about to open a PR with a Qwen3 MoE implementation I have, but happy to see it here as well (it is similar, except that it uses the RoPE implementation with complex numbers and has a state dict adapter from/to hf).
Looking over the code, I see that the MFU implementation still needs to take into account the sparse activations.
Are you interested in adding sequence packing support as well?
If not, I also have an implementation for flash attention with sequence packing using the flash_attn_varlen_func and could open a PR on it, but I am not sure if the torchtitan repo wants to stick with flex attention only.

**Comment by tianyu-l** [comment]
@jthomy 
IIUC these are two separate questions
> Looking over the code, I see that the MFU implementation still needs to take into account the sparse activations.
Are you interested in adding sequence packing support as well?

We do have sequence packing / document masking support in torchtitan using FlexAttention, which can (and should) be added to Qwen MoE. cc @wwwjn 
The MFU computation is more of following convention, lol.

> If not, I also have an implementation for flash attention with sequence packing using the flash_attn_varlen_func and could open a PR on it, but I am not sure if the torchtitan repo wants to stick with flex attention only.

There used to be discussions on this. IIRC the worry was around supporting `flash_attn_varlen_func` with CP? Even if we support `flash_attn_varlen_func`, IMO that should go into pytorch SDPA instead of into torchtitan directly.

cc @fegin @drisspg

**Comment by vwxyzjn** [comment]
@jthomy quite curious about your implementation as well. Especially the hf state dict adapter. Would you mind sharing your branch?

**Comment by jthomy** [comment]
@vwxyzjn sure, I made a draft pull request here, feel free to have a look: https://github.com/pytorch/torchtitan/pull/1688
Let me know if there's anything I can help with.

**Comment by fegin** [comment]
@jthomy FlexAttention has document masking implementation. The main blocker now is the composability with SAC, which we are working a workaround. As for SDPA version, @drisspg has prototyped a version, https://github.com/pytorch/pytorch/pull/162326/. There may be also change to how we provide these API calls due to the composability issues when enabling CP with FlexAttention. I can provide more detail later this week once I try out the proposal.

File: torchtitan/experiments/qwen3/README.md [hunk 1]
```diff
@@ -1,26 +1,32 @@
...
+## Available features
+#### Dense Model
+- Qwen3 dense model:
+    - supports FSDP/HSDP, TP, CP, DDP.
+    - Supports AC, torch.compile.
+- Qwen3 MoE model:
+    - Supports FSDP/HSDP, TP, DDP, EP.
+    - Supports AC, torch.compile.
+    - MoE models use Token Choice routing, which is using auxiluary-loss-free load balancing algorithm.
+    - [WIP] CP is not supported currently becase we used different RoPE embeding.
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/README.md]: what's our status on this? why it works on dense by not MoE? I thought RoPE only matters in the Attention layer.

When you say we used different RoPE, doesn't it mean we could've switched to (alternative but also correct) complex number based RoPE (e.g. https://github.com/pytorch/torchtitan/pull/1688) and CP would automatically work?

File: torchtitan/experiments/qwen3/README.md [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
 ```python scripts/download_hf_assets.py --repo_id <hf_repo_name> --assets tokenizer```
 
 eg, for Qwen3 0.6B model, the HF repo name is `Qwen/Qwen3-0.6B`. For 1.7B model, the HF repo name is `Qwen/Qwen3-1.7B`.
 
-#### Parity with HF
 
-Model parity test has been done and results suggest parity with HF implementation.
+## To be added
+- Dense Model
+    - Variants of Dense Models up to 32B
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/README.md]: This is referring to optimal config files to be added? I think we already have the model args for all of the dense variations.

File: torchtitan/models/moe.py [hunk 3]
```diff
@@ -231,7 +231,7 @@ def forward(
                 scores, k=self.top_k, dim=1
             )
 
-        if self.score_func == "sigmoid" and self.route_norm:
```
Comment (tianyu-l) [inline torchtitan/models/moe.py:234]: why remove this? it will change definition of the DeepSeek model

File: torchtitan/experiments/qwen3/train_configs/qwen3_moe_debug.toml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+[lr_scheduler]
+warmup_steps = 2  # lr scheduler warm up, 20% total steps
+
+[training]
+local_batch_size = 4
+seq_len = 4096
+max_norm = 1.0  # grad norm clipping
+steps = 10
+dataset = "c4"
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/train_configs/qwen3_moe_debug.toml]:
```suggestion
dataset = "c4_test"
```

File: torchtitan/experiments/qwen3/train_configs/qwen3_moe_debug.toml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+[metrics]
+log_freq = 1
+enable_tensorboard = false
+save_tb_folder = "tb"
+
+[model]
+name = "qwen3"
+flavor = "debugmodel_moe"
+hf_assets_path = "./assets/hf/Qwen3-0.6B"
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/train_configs/qwen3_moe_debug.toml]: use test tokenizer

File: torchtitan/experiments/qwen3/model/model.py [hunk 6]
```diff
@@ -281,10 +282,22 @@ def __init__(self, layer_id: int, model_args: Qwen3ModelArgs):
         self.n_heads = model_args.n_heads
         self.dim = model_args.dim
 
-        self.attention = Attention(model_args)
-        self.feed_forward = FeedForward(
-            dim=model_args.dim, hidden_dim=model_args.hidden_dim
+        self.moe_enabled = (
+            model_args.moe_enabled and (layer_id + 1) % model_args.decoder_moe_step == 0
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/model/model.py]: If `decoder_moe_step` is always 1, than it means every layer in Qwen sparse models has MoE, so we actually don't need this config.

File: torchtitan/experiments/qwen3/model/model.py [hunk 7]
```diff
@@ -281,10 +282,22 @@ def __init__(self, layer_id: int, model_args: Qwen3ModelArgs):
...
         )
+
+        self.attention = Attention(model_args)
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/model/model.py:285]: please put this line above `self.moe_enabled` definition

File: torchtitan/experiments/qwen3/README.md (hunk 1 above)
Comment (fegin) [reply torchtitan/experiments/qwen3/README.md]: ` CP is not supported` is correct but it is due to Flex?

Comment (tianyu-l) [reply torchtitan/experiments/qwen3/README.md]: @fegin oh is Flex only enabled for Qwen MoE but not Qwen dense? Either way we should update both to be consistent.

Comment (wwwjn) [reply torchtitan/experiments/qwen3/README.md]: I wrote this because of minor issue not being adressed: In Qwen, we used cos/sin RoPE embeddings, so there's no `freqs_cis` field and CP is explicitly adding freqs_cis here: https://github.com/pytorch/torchtitan/blob/71dea167b97bfdb96a538fc161a78c5669d21c51/torchtitan/train.py#L425. This is a known issue but I plan to address it later.

For Flex support - Qwen3 MoE only have one test model here, and I'm testing bigger size model. I will update bigger size MoE model to use Flex Attention. And update README to state the CP issue with FlexAttention

Comment (tianyu-l) [reply torchtitan/experiments/qwen3/README.md]: what's your plan on 
> This is a known issue but I plan to address it later.

I think I don't mind switching to `freqs_cis` based RoPE, if that makes more sense to you.

Do we know Qwen is using Flash attention of Mem Efficient attention under SDPA? @fegin 

Since (1) Flex + CP is WIP and (2) sdpa may not be the bottleneck, personally for Qwen I think it's OK to use SDPA until Flex+CP is ready. But I don't have strong opinion on this.

File: torchtitan/models/moe.py (hunk 3 above)
Comment (wwwjn) [reply torchtitan/models/moe.py:234]: Oh nice catch, yes it will change DeepSeek - I will fix that. 

In Qwen3 MoE module, https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3_moe/modeling_qwen3_moe.py#L233-L236, it used softmax + norm topk.  Should we make "route_norm" and "score function" orthogonal? 

Softmax + normalization looks a little bit redundant - The softmax function denominator already considers all samples, it looks like some kind of "normalization" to me

File: torchtitan/experiments/qwen3/model/model.py (hunk 6 above)
Comment (wwwjn) [reply torchtitan/experiments/qwen3/model/model.py]: Yes, I also noticed decoder_moe_step is always 1 in all the published model on HF. I keep the decoder_moe_step parameter to keep flexibility in experimenting interleaved moe/dense architecture (tho I'm not sure if it's really useful for model experts). I agree that we should keep it simple and remove unnecessary params.

File: torchtitan/experiments/qwen3/README.md (hunk 2 above)
Comment (wwwjn) [reply torchtitan/experiments/qwen3/README.md]: Yes we already have all different model args. This is referring to 32B training configs.

File: torchtitan/experiments/qwen3/README.md [hunk 8]
```diff
@@ -1,26 +1,29 @@
...
+## Available features
+#### Dense Model
+- Qwen3 dense model:
+    - supports FSDP/HSDP, TP, CP, DDP.
+    - Supports AC, torch.compile.
+- Qwen3 MoE model:
+    - Supports FSDP/HSDP, TP, DDP, EP.
+    - Supports AC, torch.compile.
+    - MoE models use Token Choice routing, which is using auxiluary-loss-free load balancing algorithm.
+    - [WIP] CP is not supported currently because of composability issue with FlexAttention.
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/README.md]: You are not using FlexAttention in Qwen, so this note is not accurate. Again, dense and moe should share the same CP status.

File: torchtitan/experiments/qwen3/README.md [hunk 9]
```diff
@@ -1,26 +1,29 @@
...
 
-Model parity test has been done and results suggest parity with HF implementation.
-
-#### To be added
-- Modeling
-    - Variants of Dense models up to 32B
-    - MoE alternatives
+## To be added
+- MoE model
+    - `StateDictAdapter` support for MoE model
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/README.md:26]: this is next step right?

File: torchtitan/experiments/qwen3/README.md [hunk 10]
```diff
@@ -1,26 +1,29 @@
 **The Qwen3 model is still under development.**
 
 
-#### Available features
-QWEN3 0.6B Dense model is available for:
+## Available features
+#### Dense Model
+- Qwen3 dense model:
+    - supports FSDP/HSDP, TP, CP, DDP.
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/README.md]: is CP supported? I thought because of no `freqs_cis` it won't work. What's your plan on this?

File: torchtitan/experiments/qwen3/__init__.py [hunk 11]
```diff
@@ -104,6 +106,76 @@
...
+        qk_norm=True,
+        hidden_dim=3072,
+        rope_theta=1000000,
+        moe_enabled=True,
+        moe_inter_dim=768,
+        moe_args=MoEArgs(
+            num_experts=64,
+            num_shared_experts=0,  # no shared experts, double check
+            top_k=8,  # num_experts_per_tok
+            score_func="softmax",  # need double check
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/__init__.py]: remove comments?

File: torchtitan/experiments/qwen3/__init__.py [hunk 12]
```diff
@@ -104,6 +106,76 @@
...
+            route_norm=True,
+            route_scale=1.0,  # not needed, need double check
```
Comment (tianyu-l) [inline torchtitan/experiments/qwen3/__init__.py]: remove comments?

File: torchtitan/experiments/qwen3/README.md (hunk 9 above)
Comment (wwwjn) [reply torchtitan/experiments/qwen3/README.md:26]: Yes, this is on my plate

File: torchtitan/experiments/qwen3/README.md (hunk 1 above)
Comment (fegin) [reply torchtitan/experiments/qwen3/README.md]: > @fegin oh is Flex only enabled for Qwen MoE but not Qwen dense? Either way we should update both to be consistent.

The whole model can only use one type of attention, that is the design, unless Qwen implementation explicitly allows users to configure. But I would not suggest this approach.  

Since Qwen is not using Flex, I think CP would work once the freq_cis issue @wwwjn mentioned is fixed.

**Comment by jthomy** [comment]
Did you verify that this model is the same as HF Qwen3?
E.g. I see that no scaling factor is passed to the attention, instead of `self.scaling = self.head_dim**-0.5` in HF if I am not mistaken.

**Comment by wwwjn** [comment]
> Did you verify that this model is the same as HF Qwen3? E.g. I see that no scaling factor is passed to the attention, instead of `self.scaling = self.head_dim**-0.5` in HF if I am not mistaken.

Thanks for pointing out! For dense model, yes we did some check on end-to-end forward results , eg: the description https://github.com/pytorch/torchtitan/pull/1590. But we haven't done finer-granularity checks on intermediate results. I saw you have a great test scripts in you PR and I'm thinking leveraging that!  For MoE, the numerical verification is in progress

I agree the Attention part is not the same as Qwen3 - Nice catch! And I will create a fix for that

**Comment by jthomy** [comment]
Nice, thank you! Yes feel free to us any of my code.
